lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

```diff
@@ -110,159 +245,266 @@ def get_subnet_info(subnet):
         changed = True
 
+    if module.params['tags'] != subnet['tags']:
+        if ensure_tags(conn, module, subnet, module.params['tags'], module.params['purge_tags']):
+            changed = True
+
+    subnet = get_matching_subnet(conn, module, module.params['vpc_id'], module.params['cidr'])
+
     return {
         'changed': changed,
         'subnet': subnet
     }
 
 
-def ensure_subnet_absent(conn, module, vpc_id, cidr, check_mode):
-    subnet = get_matching_subnet(conn, vpc_id, cidr)
+def ensure_subnet_absent(conn, module):
+    subnet = get_matching_subnet(conn, module, module.params['vpc_id'], module.params['cidr'])
     if subnet is None:
         return {'changed': False}
 
     try:
-        if not check_mode:
-            conn.delete_subnet(SubnetId=subnet['id'], DryRun=check_mode)
+        if not module.check_mode:
+            conn.delete_subnet(SubnetId=subnet['id'])
         return {'changed': True}
-    except botocore.exceptions.ClientError as e:
-        module.fail_json(msg=e.message, exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(e.response))
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't delete subnet")
 
 
 def main():
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            az=dict(default=None, required=False),
+            az=dict(default='', required=False),
             cidr=dict(default=None, required=True),
+            ipv6_cidr=dict(default='', required=False),
             state=dict(default='present', choices=['present', 'absent']),
             tags=dict(default={}, required=False, type='dict', aliases=['resource_tags']),
             vpc_id=dict(default=None, required=True),
-            map_public=dict(default=False, required=False, type='bool')
+            map_public=dict(default=False, required=False, type='bool'),
+            assign_instances_ipv6=dict(default=False, required=False, type='bool'),
+            wait=dict(type='bool', default=True),
+            wait_timeout=dict(type='int', default=300, required=False),
+            purge_tags=dict(default=True, type='bool')
         )
     )
 
-    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+    required_if = [('assign_instances_ipv6', True, ['ipv6_cidr'])]
 
-    if not HAS_BOTO3:
-        module.fail_json(msg='boto3 and botocore are required for this module')
+    module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=True, required_if=required_if)
 
-    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+    if module.params.get('assign_instances_ipv6') and not module.params.get('ipv6_cidr'):
+        module.fail_json(msg="assign_instances_ipv6 is True but ipv6_cidr is None or an empty string")
```

Comment:
Technically this will never be `None` because of the defaults.

Reply:
I can remove "None" from the message but you CAN cause the value to be None/null in the module by doing something stupid like this:
```yaml
    - name: create subnet
      ec2_vpc_subnet:
        cidr: "10.232.232.128/28"
        vpc_id: vpc-123456
        ipv6_cidr:
        assign_instances_ipv6: true
        state: present
```
Let me know if the preference is to change the error message.

Reply:
I thought the default was `''` so a `None` would be overridden in the default-setting?

Reply:
Argument defaults only take effect if you don't specify a parameter at all. However if you do something boneheaded and specify a parameter with no value...i.e `ipv6_cidr: ` then it won't use the default setting for the arg and you get `None` as the value.